File: melosys-eessi-app/src/test/java/no/nav/melosys/eessi/service/sed/mapper/fra_sed/melosys_eessi_melding/MelosysEessiMeldingMapperA001Test.java
package no.nav.melosys.eessi.service.sed.mapper.fra_sed.melosys_eessi_melding;

import java.util.Collections;

import no.nav.melosys.eessi.kafka.consumers.SedHendelse;
import no.nav.melosys.eessi.kafka.producers.model.MelosysEessiMelding;
import no.nav.melosys.eessi.models.SedType;
import no.nav.melosys.eessi.models.sed.SED;
import no.nav.melosys.eessi.models.sed.medlemskap.impl.MedlemskapA001;
import no.nav.melosys.eessi.models.sed.nav.Fastperiode;
import no.nav.melosys.eessi.models.sed.nav.Grunnlag;
import no.nav.melosys.eessi.models.sed.nav.Land;
import no.nav.melosys.eessi.models.sed.nav.Unntak;
import no.nav.melosys.eessi.service.joark.SakInformasjon;
import org.junit.Test;

import static no.nav.melosys.eessi.service.sed.mapper.fra_sed.melosys_eessi_melding.MelosysEessiMeldingMapperStubs.*;
import static org.assertj.core.api.Assertions.assertThat;

public class MelosysEessiMeldingMapperA001Test {

    @Test
    public void mapA001_forventRettFelt() {
        SED sed = createSed(hentMedlemskap());
        SedHendelse sedHendelse = createSedHendelse();
        SakInformasjon sakInformasjon = createSakInformasjon();
        MelosysEessiMelding melosysEessiMelding = MelosysEessiMeldingMapperFactory.getMapper(SedType.A001)
                .map("123",
                        sed,
                        sedHendelse.getRinaDokumentId(),
                        sedHendelse.getRinaSakId(),
                        sedHendelse.getSedType(),
                        sedHendelse.getBucType(),
                        sakInformasjon.getJournalpostId(),
                        sakInformasjon.getDokumentId(),
                        sakInformasjon.getGsakSaksnummer(),
                        false
                );

        assertThat(melosysEessiMelding).isNotNull();
        assertThat(melosysEessiMelding.getArtikkel()).isEqualTo("16_1");
        assertThat(melosysEessiMelding.getLovvalgsland()).isEqualTo("NO");
        assertThat(melosysEessiMelding.getAnmodningUnntak()).isNotNull();
        assertThat(melosysEessiMelding.getAnmodningUnntak().getUnntakFraLovvalgsbestemmelse()).isEqualTo("12_1");
        assertThat(melosysEessiMelding.getAnmodningUnntak().getUnntakFraLovvalgsland()).isEqualTo("SE");
    }

    private MedlemskapA001 hentMedlemskap() {
        MedlemskapA001 medlemskap = new MedlemskapA001();

        Fastperiode fastperiode = new Fastperiode();
        fastperiode.setSluttdato("2019-12-01");
        fastperiode.setStartdato("2019-05-01");
        medlemskap.setSoeknadsperiode(fastperiode);

        Land sverige = new Land();
        sverige.setLandkode("SE");
        medlemskap.setNaavaerendemedlemskap(Collections.singletonList(sverige));

        Land norge = new Land();
        norge.setLandkode("NO");
        medlemskap.setForespurtmedlemskap(Collections.singletonList(norge));

        Unntak unntak = new Unntak();
        Grunnlag grunnlag = new Grunnlag();
        grunnlag.setArtikkel("12_1");
        unntak.setGrunnlag(grunnlag);
        medlemskap.setUnntak(unntak);

        return medlemskap;
    }
}

File: melosys-eessi-app/src/test/java/no/nav/melosys/eessi/service/sed/mapper/fra_sed/melosys_eessi_melding/MelosysEessiMeldingMapperA001Test.kt
package no.nav.melosys.eessi.service.sed.mapper.fra_sed.melosys_eessi_melding

import io.kotest.matchers.nulls.shouldNotBeNull
import io.kotest.matchers.shouldBe
import io.kotest.matchers.shouldNotBe
import no.nav.melosys.eessi.models.SedType
import no.nav.melosys.eessi.models.sed.medlemskap.impl.*
import no.nav.melosys.eessi.models.sed.nav.Fastperiode
import no.nav.melosys.eessi.models.sed.nav.Grunnlag
import no.nav.melosys.eessi.models.sed.nav.Land
import no.nav.melosys.eessi.models.sed.nav.Unntak
import no.nav.melosys.eessi.service.sed.mapper.fra_sed.melosys_eessi_melding.MelosysEessiMeldingMapperStubs.createSed
import no.nav.melosys.eessi.service.sed.mapper.fra_sed.melosys_eessi_melding.MelosysEessiMeldingMapperStubs.createSedHendelse
import org.junit.jupiter.api.Test

class MelosysEessiMeldingMapperA001Test {

    private val melosysEessiMeldingMapperFactory = MelosysEessiMeldingMapperFactory("dummy")

    private fun mapMedlemskap(medlemskap: MedlemskapA001) =
        createSedHendelse().let { hendelse ->
            melosysEessiMeldingMapperFactory.getMapper(SedType.A001).map(
                EessiMeldingParams(
                    aktoerId = "123",
                    sed = createSed(medlemskap),
                    rinaDokumentID = hendelse.rinaDokumentId,
                    rinaSaksnummer = hendelse.rinaSakId,
                    sedType = hendelse.sedType,
                    bucType = hendelse.bucType,
                    avsenderID = hendelse.avsenderId,
                    landkode = "landkode",
                    sedErEndring = false,
                    sedVersjon = "1"
                )
            )
        }

    @Test
    fun mapA001_forventRettFelt() {
        val melosysEessiMelding = mapMedlemskap(hentMedlemskapCdm43())

        melosysEessiMelding.shouldNotBeNull().run {
            artikkel shouldBe "16_1"
            lovvalgsland shouldBe "NO"
            anmodningUnntak shouldNotBe null
            anmodningUnntak?.unntakFraLovvalgsbestemmelse shouldBe "12_1"
            anmodningUnntak?.unntakFraLovvalgsland shouldBe "SE"
        }
    }

    @Test
    fun `mapA001 CDM 4_4 med forordning8832004 grunnlag`() {
        val melosysEessiMelding = mapMedlemskap(hentMedlemskapCdm44())

        melosysEessiMelding.shouldNotBeNull().run {
            artikkel shouldBe "16_1"
            lovvalgsland shouldBe "NO"
            anmodningUnntak.shouldNotBeNull().run {
                unntakFraLovvalgsbestemmelse shouldBe "16_1"
                unntakFraLovvalgsland shouldBe "SE"
                erFjernarbeidTWFA shouldBe false
            }
        }
    }

    @Test
    fun `mapA001 CDM 4_4 med TWFA markering`() {
        val medlemskap = hentMedlemskapCdm44().apply {
            rammeavtale = Rammeavtale(fjernarbeid = Fjernarbeid(eessiYesNoType = "ja"))
        }

        val melosysEessiMelding = mapMedlemskap(medlemskap)

        melosysEessiMelding.shouldNotBeNull().run {
            anmodningUnntak.shouldNotBeNull().run {
                erFjernarbeidTWFA shouldBe true
            }
        }
    }

    @Test
    fun `mapA001 CDM 4_3 fallback til unntak grunnlag`() {
        val medlemskap = hentMedlemskapCdm43()

        val melosysEessiMelding = mapMedlemskap(medlemskap)

        melosysEessiMelding.shouldNotBeNull().run {
            anmodningUnntak.shouldNotBeNull().run {
                unntakFraLovvalgsbestemmelse shouldBe "12_1"
                erFjernarbeidTWFA shouldBe false
            }
        }
    }

    @Test
    fun `mapA001 CDM 4_3 uten TWFA`() {
        val melosysEessiMelding = mapMedlemskap(hentMedlemskapCdm43())

        melosysEessiMelding.shouldNotBeNull().run {
            anmodningUnntak.shouldNotBeNull().run {
                erFjernarbeidTWFA shouldBe false
            }
        }
    }

    private fun hentMedlemskapCdm43(): MedlemskapA001 = MedlemskapA001().apply {
        soeknadsperiode = Fastperiode().apply {
            sluttdato = "2019-12-01"
            startdato = "2019-05-01"
        }
        naavaerendemedlemskap = mutableListOf(Land().apply { landkode = "SE" })
        forespurtmedlemskap = mutableListOf(Land().apply { landkode = "NO" })
        unntak = Unntak().apply {
            grunnlag = Grunnlag().apply { artikkel = "12_1" }
        }
    }

    private fun hentMedlemskapCdm44(): MedlemskapA001 = MedlemskapA001().apply {
        soeknadsperiode = Fastperiode().apply {
            sluttdato = "2019-12-01"
            startdato = "2019-05-01"
        }
        naavaerendemedlemskap = mutableListOf(Land().apply { landkode = "SE" })
        forespurtmedlemskap = mutableListOf(Land().apply { landkode = "NO" })
        unntak = Unntak().apply {
            begrunnelse = "Begrunnelse"
        }
        forordning8832004 = Forordning8832004(
            unntak = UnntakForordning(
                grunnlag = Grunnlag().apply { artikkel = "16_1" }
            )
        )
    }
}
